From: Chad Dupuis <chad.dupuis@qlogic.com>
Date: Mon, 11 Aug 2014 06:20:09 -0400
Subject: qla2xxx: Do not decrement the reference count in abort handler if only one reference is held.
References: bnc#891212
Patch-Mainline: not yet

If a command cannot be found in firmware and the abort handler is the
only entity holding a reference to the command, don't decrement the
reference counter as we will hit a BUG_ON when try to free the SRB.

Signed-off-by: Chad Dupuis <chad.dupuis@qlogic.com>
Acked-by: Hannes Reinecke <hare@suse.de>
---
 drivers/scsi/qla2xxx/qla_os.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/scsi/qla2xxx/qla_os.c b/drivers/scsi/qla2xxx/qla_os.c
index 67b64b9..9588e72 100644
--- a/drivers/scsi/qla2xxx/qla_os.c
+++ b/drivers/scsi/qla2xxx/qla_os.c
@@ -968,9 +968,11 @@ qla2xxx_eh_abort(struct scsi_cmnd *cmd)
 		if (rval == QLA_FUNCTION_PARAMETER_ERROR) {
 			/*
 			 * Decrement the ref_count since we can't find the
-			 * command
+			 * command if there is another reference to the
+			 * command.
 			 */
-			atomic_dec(&sp->ref_count);
+			if (atomic_read(&sp->ref_count) > 1)
+				atomic_dec(&sp->ref_count);
 			ret = SUCCESS;
 		} else
 			ret = FAILED;
-- 
1.8.5.2

